agent/cache-types/connect_ca_leaf.go

```diff
@@ -401,10 +401,20 @@ func (c *ConnectCALeaf) Fetch(opts cache.FetchOptions, req cache.Request) (cache
 	c.fetchStart(rootUpdateCh)
 	defer c.fetchDone(rootUpdateCh)
 
-	// Setup the timeout chan outside the loop so we don't keep bumping the timout
+	// Setup the timeout chan outside the loop so we don't keep bumping the timeout
 	// later if we loop around.
 	timeoutCh := time.After(opts.Timeout)
 
+	// If we called Fetch() with MustRevalidate then this call came from a non-blocking query.
+	// To pickup any CA rotations that may have happened prior to this call, we will reduce the timeout
+	// to ~ 1 sec. Otherwise, we risk non-blocking queries taking ~ 10 minutes to wait on a CA rotation
+	// that could never happen.
+	// see https://github.com/hashicorp/consul/issues/10871, https://github.com/hashicorp/consul/issues/9862
+	// This is a hack, so, ideally, we can revisit the caching/ fetching logic in this case
+	if req.CacheInfo().MustRevalidate {
+		timeoutCh = time.After(time.Second)
+	}
```

Comment:
I think this works, but I also wonder if we could make the non-blocking case a bit more obvious by avoiding the select entirely.

What if we were to:
* move this `if` above where we create `rootUpdateCh` around line 382
* remove the second `if` below (we'll handle it in this same if block)
* instad of `timeoutCh = time.After(time.Second)` we do something like this:

```go
roots, err := c.rootsFromCache()
if err != nil {
    return lastResultWithNewState(), err
}
if activeRootHasKey(roots, state.authorityKeyID) {
        return lastResultWithNewState(), err
}
... // some shared logic here to calculate expiresAt
if state.forceExpireAfter.Before(expiresAt) {
    return c.generateNewLeaf(reqReal, lastResultWithNewState())
}
return lastResultWithNewState(), err
```

The `...` about calculating `expiresAt` could be extracted from the 461-470 into a new function and called from both places.

I haven't fully gone through that to be sure that works, but my impression is that something like that would work and allow us to avoid the `select` which makes it a bit more obvious that we are handling  a non-blocking request, without having to set a short timeout.

Reply:
Hey thanks for bringing this up again -- I recall we chatted about this approach and dropped the ball as I was debugging the "vault issue".  💯  

I think it makes a lot of sense (readability, maintainability) to treat the `MustRevalidate`/ non-blocking case as "separate" from the select stmt. 

Immediately after we check : https://github.com/hashicorp/consul/blob/ed8a901be72a45d28b70ea357c44e0b009a83cb0/agent/cache-types/connect_ca_leaf.go#L378-L381

If we add something like ~:

```
	if req.CacheInfo().MustRevalidate {
		roots, err := c.rootsFromCache()
		if err != nil {
			return lastResultWithNewState(), err
		}
		if activeRootHasKey(roots, state.authorityKeyID) {
			return lastResultWithNewState(), nil
		}

		// if we reach here then the current leaf was not signed by the same CAs, just regen
		return c.generateNewLeaf(reqReal, lastResultWithNewState())
	}
```

That will catch most occurrences of https://github.com/hashicorp/consul/issues/10871. Sometimes, the CA may take a bit to be rotated but that's ok. "Certainty" / " strong consistency" is what the blocking queries are for.

Also note, that we don't need to "calculate" when to regen the cert. If for whatever reason we don't have a CA that signed the leaf, we need to regen the leaf cert.

Wdyt?